all-actors/src/main/java/org/sunbird/SvgGenerator.java

```diff
@@ -0,0 +1,148 @@
+
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.IOException;
+import java.io.StringWriter;
+import java.io.UnsupportedEncodingException;
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.net.URLEncoder;
+import java.nio.charset.StandardCharsets;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Set;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+public class SvgGenerator {
+
+    private Logger logger = LoggerFactory.getLogger(SvgGenerator.class);
+    private String svgTemplate;
+    private String directory;
+    private static Map<String, String> encoderMap = new HashMap<>();
+
+
+    static {
+        Velocity.setProperty(RuntimeConstants.RESOURCE_LOADER, "classpath");
+        Velocity.setProperty("classpath.resource.loader.class", ClasspathResourceLoader.class.getName());
+        Velocity.init();
+        encoderMap.put("<", "%3C");
+        encoderMap.put(">", "%3E");
+        encoderMap.put("#", "%23");
+        encoderMap.put("%", "%25");
+        encoderMap.put("\"", "\'");
+    }
+
+    public SvgGenerator(String svgTemplate, String directory) {
+        this.svgTemplate = svgTemplate;
+        this.directory = directory;
+    }
+
+    public String generate(CertificateExtension certificateExtension, String encodedQrCode) {
+        String svgFileName = getSvgFileName();
+        String svgContent;
+        File file = new File(directory + svgFileName);
+        if (!file.exists()) {
+            download(svgFileName);
+        }
+        svgContent = readSvgContent(file.getAbsolutePath());
+        String encodedSvg = null;
+        try {
+            encodedSvg = "data:image/svg+xml," + replaceTemplateVars(encodeData(svgContent), certificateExtension, encodedQrCode);
+            encodedSvg = encodedSvg.replaceAll("\n", "").replaceAll("\t", "");
+        } catch (IOException e) {
+            logger.info("SvgGenerator:generate exception while encoding svgContent {}", e.getMessage());
+        }
+        return encodedSvg;
+    }
+
+
+    private String replaceTemplateVars(String svgContent, CertificateExtension certificateExtension, String encodeQrCode) throws IOException {
+        HTMLVarResolver htmlVarResolver = new HTMLVarResolver(certificateExtension);
+        VelocityContext context = new VelocityContext();
+        Set<String> htmlReferenceVariable = HTMLTemplateValidator.storeAllHTMLTemplateVariables(svgContent);
+        for (String var : htmlReferenceVariable) {
+            String macro = var.substring(1);
+            try {
+                Method method = htmlVarResolver.getClass().getMethod("get" + StringUtils.capitalize(macro));
+                method.setAccessible(true);
+                String data = URLEncoder.encode((String) method.invoke(htmlVarResolver), "UTF-8");
+                context.put(macro, data.replace("+", "%20"));
```

Comment:
Pls put an explicit comment here that the URLEncoder.encode does not replace +. This provides a learning opportunity to fellow developers.

Reply:
done